Introduce ExternalSessionService.

demiankatz opened the pull request:
This migrates most of the external session handling logic to a new database service. The expiration cleanup command can be ported after #3714 is merged, since that work requires changes pending in that other PR.

All tests are passing, but I don't have a real-world environment to test this with, so help with that would be appreciated!

demiankatz committed:
Introduce ExternalSessionService.

EreMaijala commented on module/VuFind/src/VuFind/Db/Table/ExternalSession.php:
```diff
@@ -103,10 +107,12 @@ public function getByExternalSessionId($sid)
      *
      * @return void
+     *
+     * @deprecated Use ExternalSessionServiceInterface::addSessionMapping()
```
This should point to `ExternalSessionServiceInterface::destroySession()`

EreMaijala commented on module/VuFind/src/VuFind/Db/Table/ExternalSession.php:
```diff
@@ -91,10 +93,12 @@ public function addSessionMapping($localSessionId, $externalSessionId)
      *
      * @return ?\VuFind\Db\Row\ExternalSession
+     *
+     * @deprecated Use ExternalSessionServiceInterface::addSessionMapping()
```
Wrong method name here.

EreMaijala commented on module/VuFind/src/VuFind/Db/Service/ExternalSessionServiceInterface.php:
```diff
@@ -0,0 +1,82 @@
+     *
+     * @return ?ExternalSessionEntityInterface
+     */
+    public function getExternalSessionById(string $sid): ?ExternalSessionEntityInterface;
```
I believe the old name was intentionally get**By**ExternalSessionId so as to not confuse it with the table id column.

EreMaijala commented:
@demiankatz I'll see if I can add a Mink test for this using SimulatedSSO.

demiankatz committed:
Merge branch 'dev' into external-session-service

demiankatz committed:
Review feedback.

demiankatz committed:
Style.

demiankatz commented:
Thanks for the review, @EreMaijala -- I've addressed your comments (including changing getExternalSessionById to getExternalSessionByExternalSessionId, which is long and redundant but more clear!).

Thanks also for the offer of test coverage -- more never hurts, though it does appear that we're at least using some parts of the service in ShibbolethLogoutNotificationTest.

EreMaijala commented:
@demiankatz Indeed, but I think it's still worth adding it to SimulatedSSO test so that it's the "real thing" adding the session mapping.

demiankatz commented:
I'm currently working on adding the expiration command...

EreMaijala committed:
Add external logout test with SimulatedSSO.

demiankatz committed:
Add expiration command (includes some fixes to related AuthHash code).

demiankatz committed:
php-cs-fixer

demiankatz committed:
Shorten method name.
